### DISCUSSION TITLE
Dudas varias sobre material de Apuntes

### DISCUSSION THREAD

**Discussion by heinHertz** [dds-jv/foro#171] [Consultas de objetos] [answered]
Tengo varias dudas para repasar:
1) como es una implementación correcta de un Builder?  es decir , en  el libro de [Patrones de Diseño](https://www.utnianos.com.ar/foro/attachment.php?aid=3577)  en la parte del Patron builder hay un Director, yo tenia entendido que una implementación es que el usuario del sistema ingresa manualmente lo que necesita, por ejemplo:
Barista builder2 = new Barista()
Builder.conCafe()
Bebida bebidaBruno = builder2.crearBebida()
que hace ese Director? en realidad estos métodos tienen que estar dentro de ese Director? 

2)  no entiendo el code smell Data Classes , del apunte [Manejo del Cambio y Refactor](https://docs.google.com/document/d/1cAje0qwy3Cus_ob0r-tatbcT01sDFeLt3MmSVmLeSxk/edit#)
"Una clase que sólo representa una estructura de datos, es fácil reconocerla cuando tiene sólo getters y setters. "
A) esta hablando de los repositorios? quizá mi problema es distinguir singleton de repositorio, lo que entiendo es esto:
singleton tiene una unica instancia, o sea  hay una lista sola en todo ciclo de funcionamiento, para todos y repositorio es una clase con una lista ?  no es justamente un singleton o un repo una estructura de datos?

B) y tambien habla de una mala practica de separar un atributo de sus métodos ?
en realidad esta hablando de una clase con solamente atributos y sin metodos?

3) no entiendo que hace un metodo estatico? se que el atributo de clase cambia para todas las instancias de la Clases que contengan a ese Static , pero no que hace exactamente el metodo static  y para que se puede usar.

**Comment by gastonprieto** [comment]
Buenas @heinHertz !

1. Recorda que lo que define el libro son responsabilidades, no necesariamente clases. En este caso el director es ese pseudocódigo que pusiste.

```
Barista builder2 = new Barista()
Builder.conCafe()
Bebida bebidaBruno = builder2.crearBebida()
```

Que seguramente este en la UI o en un test, entonces esos serían los directores en este caso. En otros casos podría ser otro objeto.

**Comment by gastonprieto** [comment]
2a.  Singleton y repositorio son dos conceptos distintos. El singleton habla del aspecto creacional de un objeto, en este caso de que se va a construir uno y que se va a acceder de forma global.

Repositorio habla de un objeto que tiene la responsabilidad de conocer a todas las instancias de cierto tipo y devolvertelas todas o algunas filtradas bajo algún criterio. 

Un repositorio puede ser Singleton? Si. Puede no serlo? Si, tambien. En este caso que nuestros repositorios guardan las instancias que conocen en memoria si necesitas que sean Singleton, porque quisieras que todos usen el mismo, sino van a estar perdiendo las instancias. Ahora si tuvieras que conseguir la información de una base de datos o de un archivo no te molestaría tener muchas instancias, porque tu memoria compartida serían los archivos/base de datos.

Se entendió un poco mas la diferencia?

**Comment by gastonprieto** [comment]
> en realidad esta hablando de una clase con solamente atributos y sin metodos?

Si. O sea sin comportamiento interesante. Por ejemplo un error sería hacer

```Java
class Prenda {
   int costo;
   //setters/gettters
}

class CalculadorPrecioPrenda {
   public float precio(prenda) return prenda.costo() * 1.1
}
```

En este caso estas separando los datos de su comportamiento. Que es basicamente la ventaja de tener objetos, que es de tener entidades que puedan agrupar comportamiento y estado en un mismo lugar.

**Comment by gastonprieto** [comment]
> no entiendo que hace un metodo estatico? se que el atributo de clase cambia para todas las instancias de la Clases que contengan a ese Static , pero no que hace exactamente el metodo static y para que se puede usar.

Es un mensaje que le mandas a la clase, o sea, no necesitas instanciar un objeto para poder mandarle el mensaje. 
Pensa sino como harías para conocer la única instancia de un Singleton, si necesitas instanciar un objeto nuevo, para poder obtener el único que existe, entonces ya estarías en un problema porque tendrías dos. 

El problema que tienen estos mensajes es que no podrías tener dos clases que entiendan el mismo mensaje y sean polimorficas, no se heredan tampoco y son dificiles de mockear.

**Comment by heinHertz** [comment, marked as answer]
me sirvió, gracias Gastón !
